spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerTemplate.java

```diff
@@ -0,0 +1,147 @@
+
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+
+/**
+ * Offers convenience methods for performing common operations on Secret Manager including
+ * creating and reading secrets.
+ *
+ * @author Daniel Zou
+ * @since 1.3
+ */
+public class SecretManagerTemplate implements SecretManagerOperations {
+
+	private final SecretManagerServiceClient secretManagerServiceClient;
+
+	private final GcpProjectIdProvider projectIdProvider;
+
+	public SecretManagerTemplate(
+			SecretManagerServiceClient secretManagerServiceClient,
+			GcpProjectIdProvider projectIdProvider) {
+		this.secretManagerServiceClient = secretManagerServiceClient;
+		this.projectIdProvider = projectIdProvider;
+	}
+
+	@Override
+	public void createSecret(String secretId, String payload) {
+		if (!secretExists(secretId)) {
+			createSecret(secretId);
+		}
+
+		createNewSecretVersion(secretId, ByteString.copyFromUtf8(payload));
+	}
+
+	@Override
+	public void createSecret(String secretId, byte[] payload) {
+		if (!secretExists(secretId)) {
+			createSecret(secretId);
+		}
+
+		createNewSecretVersion(secretId, ByteString.copyFrom(payload));
+	}
+
+	@Override
+	public String getSecretString(String secretId, String versionName) {
+		return getSecretByteString(secretId, versionName).toStringUtf8();
+	}
+
+	@Override
+	public byte[] getSecretBytes(String secretId, String versionName) {
+		return getSecretByteString(secretId, versionName).toByteArray();
+	}
+
+	@Override
+	public ByteString getSecretByteString(String secretId, String versionName) {
+		SecretVersionName secretVersionName = SecretVersionName.of(
+				this.projectIdProvider.getProjectId(),
+				secretId,
+				versionName);
+
+		AccessSecretVersionResponse response = secretManagerServiceClient.accessSecretVersion(secretVersionName);
+
+		return response.getPayload().getData();
+	}
+
+	/**
+	 * Create a new version of the secret with the specified payload under a {@link Secret}.
+	 */
+	private void createNewSecretVersion(String secretId, ByteString byteStringPayload) {
+		SecretName name = SecretName.of(projectIdProvider.getProjectId(), secretId);
+		SecretPayload payloadObject = SecretPayload.newBuilder()
+				.setData(byteStringPayload)
+				.build();
+
+		AddSecretVersionRequest payloadRequest = AddSecretVersionRequest.newBuilder()
+				.setParent(name.toString())
+				.setPayload(payloadObject)
+				.build();
+		secretManagerServiceClient.addSecretVersion(payloadRequest);
+	}
+
+	/**
+	 * Creates a new secret for the GCP Project.
+	 *
+	 * <p>
+	 * Note that the {@link Secret} object does not contain the secret payload. You must
+	 * create versions of the secret which stores the payload of the secret.
+	 */
+	private void createSecret(String secretId) {
+		ProjectName projectName = ProjectName.of(projectIdProvider.getProjectId());
+
+		Secret secret = Secret.newBuilder()
+				.setReplication(
+						Replication.newBuilder()
+								.setAutomatic(Replication.Automatic.newBuilder().build())
```

Comment:
You can remove the intermediate `build()` calls -- proto builders usually take ether a final object or its builder.

Reply:
Done.